Step3 질문삭제기능 리팩토링

mmtos opened the pull request:
[소감]
응용서비스에서 도메인 객체로 로직들을 이동하니 repository와 같은 복잡한 의존관계를 끌어들이지 않아도 테스트케이스 작성을 할수 있게되어 편했습니다. 

[질문]
1. 하나의 도메인에 여러개의 엔티티(Question, Answer, DeleteHistory)가 포함되도 괜찮을까요?  
  - 이번에 읽은 책[최범균 - 도메인 주도 개발 시작하기]에는 하나의 도메인에는 보통 하나의 엔티티만 있게된다고 언급이 되어있어서 궁금해졌습니다.

mmtos committed:
docs : 요구사항 정리 및 도메인 분석

mmtos committed:
refactor: 도메인 엔티티만 테스트하기위해 클래스 분리

mmtos committed:
feat: 질문 삭제 기능 이동
- 로그인 사용자와 질문한 사람이 동일한 경우 삭제 가능
- 로그인 사용자와 질문한 사람이 동일하지 않은 경우 삭제 불가

mmtos committed:
refactor: 테스트케이스 리팩토링

mmtos committed:
feat: 질문 삭제 기능 리팩토링
- 답변이 없는 경우 삭제 가능
- 질문자와 모든 답변 글의 답변자가 동일한 경우 삭제 가능
- 질문자와 답변자가 다른 경우 삭제 불가

mmtos committed:
test: 테스트 assert 추가 및 리팩토링
- Question 삭제시 삭제히스토리 검증 추가
- QnaService 역할에 맞게 테스트케이스 리팩토링
- 필요없는 테스트 클래스 제거
- 테스트를 위해 package레벨의 생성자 추가(Answer, Question)

joojimin commented on docs/QnA.md:
```diff
@@ -29,6 +52,14 @@
+   - [X] 로그인 사용자와 질문한 사람이 동일하지 않은 경우 삭제 불가
+   - [X] 답변이 없는 경우 삭제 가능
+   - [X] 질문자와 모든 답변 글의 답변자가 동일한 경우 삭제 가능
+   - [X] 질문자와 답변자가 다른 경우 삭제 불가
```
요구사항 정리 👍 👍

joojimin replied:
요구사항 정리 👍 👍

joojimin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -35,6 +37,16 @@ protected Answer() {
+        this.deleted = deleted;
+    }
+
+    Answer(Long id, User writer, Question question, String contents) {
```
테스트용으로만 쓰이는 생성자인거같은데 id를 입력받는 생성자가 꼭 필요할까요~?

joojimin replied:
테스트용으로만 쓰이는 생성자인거같은데 id를 입력받는 생성자가 꼭 필요할까요~?

joojimin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -49,6 +61,18 @@ public Answer(User writer, Question question, String contents) {
+    DeleteHistory delete(User loginUser) {
+        verifyWriter(loginUser);
+        this.deleted = true;
+        return new DeleteHistory(ContentType.ANSWER,this.id, loginUser, LocalDateTime.now());
```
`new DeleteHistory(ContentType.ANSWER,this.id, loginUser, LocalDateTime.now());` DeleteHistory를 만드는 정적팩토리메서드를 만들어보면 어떨까요~? ContentType의 Enum값으로 구분되기에 잘못된 값이 셋팅될 실수를 줄일수 있을것같습니다!

추가적으로 질문해주신것과도 연관있을거같은데
Answer에서 DeleteHistory 도메인을 만드는 책임까지 가지고있는 것 같아요!
같은 에그리거트를 갖는 범위라면 괜찮을수도있지만 도메인끼리의 경계가 명확한 것이 더 좋을것같습니다!

joojimin replied:
`new DeleteHistory(ContentType.ANSWER,this.id, loginUser, LocalDateTime.now());` DeleteHistory를 만드는 정적팩토리메서드를 만들어보면 어떨까요~? ContentType의 Enum값으로 구분되기에 잘못된 값이 셋팅될 실수를 줄일수 있을것같습니다!

추가적으로 질문해주신것과도 연관있을거같은데
Answer에서 DeleteHistory 도메인을 만드는 책임까지 가지고있는 것 같아요!
같은 에그리거트를 갖는 범위라면 괜찮을수도있지만 도메인끼리의 경계가 명확한 것이 더 좋을것같습니다!

joojimin reviewed (changes requested):
안녕하세요! 승훈님!
Step3 미션도 고생하셧습니다~
몇가지 생각해볼 피드백 드렸습니다! 확인후 리뷰재요청부탁드려요~

mmtos committed:
review-feedback: id를 받는 기본 생성자 삭제(Question, Answer)

mmtos committed:
review-feedback: DeleteHistoryFactory 작성

mmtos committed:
fix: QuestionTest 오류 수정
- 메서드 동작과 일치하지 않는 메서드명 수정
- 테스트케이스를 제대로 반영하지 못하는 테스트 메서드 수정

mmtos committed:
refactor: 질문삭제시도부분 메서드화
- 테스트 작성시 실수한 부분에 대해 앞으로 헷갈리지 않기한 목적

mmtos committed:
review-feedback: QuestionDeleteService 작성
- DeleteHistory 생성 책임을 Question, Answer에서 분리
- QuestionDeleteServiceTest 작성
- QuestionTest 리팩토링

mmtos committed:
format: java code style 적용

mmtos commented:
저번 테스트 구현에서 id를 포함한 생성자를 사용했던 이유는, 
  - 삭제히스토리 생성을 위한 id가 필요했고,
  - 실제 id를 얻기위해선 DB접근 필수였는데(GenerationType = IDENTITY)
  - 하지만 JPA 의존성없이 도메인 레이어 테스트를 해야했기때문입니다. 

이번 리뷰를 보고 다른 방법(Mock, Stub 등등)은 없나 고민해봤습니다.  JPA에서는 어떻게 id 필드에 값을 주입하지?라는 의문이 들어서 찾아보니 Reflection API를 사용해서 주입하는것 같더라구요.   
그래서 저도 테스트코드에서만 Reflection API를 사용해서 Id값을 주입하는 방식으로 리팩토링 해보았습니다! 
혹시 다른 방법도 있다면 소개 부탁드려요

이번 리뷰도 잘 부탁드립니다. ㅎㅎ

joojimin reviewed (approved):
고생하셧습니다! 승훈님~
조금 빡빡한 리뷰를 드렸는데 새로운 방법을 고민하시고 적용해주셧네요!!
이번미션은 여기서 머지하겟습니다!
남은 미션도 화이팅입니다!
